feat: add documentation regarding using Job Bundles as an integration (#13)
aws-deadline/deadline-cloud-samples: job_bundles/README.md

[configuring the Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud/blob/mainline/README.md#configuration) 
you can run it with the Blender sample in this samples directory (`deadline bundle gui-submit blender_render/`) to see:

![UI Shared Settings](../.images/blender_submit_shared_settings.png) ![UI Job Settings](../.images/blender_submit_job_settings.png) ![UI Job Attachments](../.images/blender_submit_job_attachments.png) 

where the contents of the Job-specific settings panel are automatically generated using the `userInterface` properties of Job Parameters
[defined in the Job Template](https://github.com/aws-deadline/deadline-cloud-samples/blob/bdd5ff5ea29eb7457c9a78ba39166b891b79151e/job_bundles/blender_render/template.yaml#L11-L19) within 
your Job Bundle, and the [Queue Environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html)
defined on the Queue that you are submitting to.

All of the Job submitter plugins that have been developed by the AWS Deadline Cloud team, such as the 
[Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your

> Job submission then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) 

Reviewer: Need a comma for grammar
```diff
-Job submission then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) 
+Job submission then, using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) 
```

Author: Thanks for drawing attention to this.

submit your Job to Deadline Cloud. You can see the Job Bundles that are
submitted by looking in the job history directory on your workstation after submitting a Job using a submitter plugin or the Deadline
Cloud CLI. You can find your job history directory by running the command: `deadline config get settings.job_history_dir`.

## Elements of a Job Bundle

A Job Bundle is a directory structure that contains at least an
[Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) file, but may contain
other files as follows:

```
/template.json (or template.yaml)
/asset_references.json (or asset_references.yaml)
/parameter_values.json (or parameter_values.yaml)
<plus any other Job-specific files that you'd like>
```

The only required file is the Job Template (`template.json`/`template.yaml`) file that describes the structure and behaviour

> of your Job. The remaining files are optional, and are described in the following subsections.

Reviewer: Perhaps slightly different wording?
```diff
-of your Job. The remaining files are optional, and are described in the following subsections.
+of your Job. The other files are optional, and are described in the following subsections.
```

### Elements - Job Template

The `template.json`/`template.yaml` file defines the runtime environment and all the processes that will run
as part of an Amazon Deadline Cloud Job. It can be parameterized so that the same template can be used to
create Jobs that differ only in their input values; much like a function or template in your favourite programming
langauge. 

For example, the [Job Template for the `blender_render` sample](https://github.com/aws-deadline/deadline-cloud-samples/blob/mainline/job_bundles/blender_render/template.yaml)
defines input parameters like `BlenderSceneFile` which is a file path:

```yaml
- name: BlenderSceneFile
  type: PATH
  objectType: FILE
  dataFlow: IN
  userInterface:
    control: CHOOSE_INPUT_FILE
    label: Blender Scene File
    groupLabel: Render Parameters
    fileFilters:
    - label: Blender Scene Files
      patterns: ["*.blend"]
    - label: All Files
      patterns: ["*"]
  description: >
    Choose the Blender scene file you want to render. Use the 'Job Attachments' tab
    to add textures and other files that the job needs.
```

> The Job Parameter properties `userInterface`, `objectType`, and `dataFlow` are used by the Deadline Cloud CLI when

Reviewer: Sorry nitpicking
```diff
-The Job Parameter properties `userInterface`, `objectType`, and `dataFlow` are used by the Deadline Cloud CLI when
+ `dataFlow`, `objectType` and `userInterface` are optional Job Parameter properties used by the Deadline Cloud CLI when
```

present in a Job Template within a Job Bundle. 

> The `userInterface` properties control the behaviour of automatically generated Job submission UIs; both via 

Reviewer: ```diff
-The `userInterface` properties control the behaviour of automatically generated Job submission UIs; both via 
+The `userInterface` property controls the behaviour of automatically generated Job submission UIs; both via 
```

the `deadline bundle gui-submit` command line, and within Job submittion plugins for applications
such as the [Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya). 
In this example, the UI widget for inputting a value for `BlenderSceneFile` will be a file-selection dialog that

> allows filtering to see only Blender's `.blend` files or all files, and within a widget group called "Render Parameters":

Reviewer 2: suggest: add a link to the GUI control showcase sample for other types of userInterface examples that people can copypasta. https://github.com/aws-deadline/deadline-cloud-samples/tree/mainline/job_bundles/gui_control_showcase

![BlenderSceneFile UI Widget](../.images/blender_submit_scene_file_widget.png)

The `objectType` and `dataFlow` properties control the behaviour of Deadline Cloud's 
[Job Attachments feature](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html) when submitting
a Job from the Job Bundle. In this case, `objectType: FILE` and `dataFlow: IN` mean that the value of `BlenderSceneFile`
will be treated as an input file for Job Attachments. Contrast that with the definition of the `OutputDir` which has
`objectType: DIRECTORY` and `dataFlow: OUT`:

```yaml
- name: OutputDir
  type: PATH
  objectType: DIRECTORY
  dataFlow: OUT
  userInterface:
    control: CHOOSE_DIRECTORY
    label: Output Directory
    groupLabel: Render Parameters
  default: "./output"
  description: Choose the render output directory.
```

> The value of `OutputDir` is treated by Job Attachments as a directory where the Job is expected to write output files.

Reviewer: Include a link out to OpenJD documentation about where I can learn more about these properties or what the values can be for them?

Reviewer 3: Question - do we need to include any of the Worker Agent execution aspects such as file & folder permissions as expectations. Yes I know its not part of the bundle spec, but it does set an expectation of crafting the bundle, and likely and scripts defined in OpenJD steps.

Author: > Question - do we need to include any of the Worker Agent execution aspects such as file & folder permissions as expectations. Yes I know its not part of the bundle spec, but it does set an expectation of crafting the bundle, and likely and scripts defined in OpenJD steps.

I'm picturing a separate document where we talk about path mapping, storage profiles, and job attachments in more depth. I think that this would go in there. I'd like to keep this narrowly focused on just the Job Bundle; it's very easy to go down a path here where we describe all of the nuances of job attachments in here as well.

> The remainder of the `blender_render` sample's Job Template defines the workflow as a single Step with each frame of the animation

Reviewer: Perhaps a link out to OpenJD docs about this, or at least explaining the `embeddedFiles` perhaps?

          # Configure the task to fail if any individual command fails.
          set -xeuo pipefail

          mkdir -p '{{Param.OutputDir}}'

          blender --background '{{Param.BlenderSceneFile}}' \
                  --render-output '{{Param.OutputDir}}/{{Param.OutputPattern}}' \
                  --render-format {{Param.Format}} \
                  --use-extension 1 \
                  --render-frame {{Task.Param.Frame}}
```

Additional samples of Job Templates can be found in all of the Job Bundles in this repository as well as in the 
[samples provided](https://github.com/OpenJobDescription/openjd-specifications/tree/mainline/samples)
by Open Job Description. Please see https://github.com/OpenJobDescription/openjd-specifications/wiki
for more information on Job Templates.

### Elements - Parameter Values

The `parameter_values.json`/`parameter_values.yaml` file in a Job Bundle gives you a place to "bake"
the values of some of the Job Parameters, and/or [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html)
request arguments into the Job Bundle so that the values do not have to be given when submitting a Job using the

> Job Bundle. The Job Bundle submission UI that is created will still allow values for these parameters to be input.

Reviewer: ```diff
-Job Bundle. The Job Bundle submission UI that is created will still allow values for these parameters to be input.
+Job Bundle. The Job Bundle submission UI that is created will still allow values for these parameters to be modified.
```

The format of the file, in YAML, is:

```yaml
parameterValues:
- name: <string>
  value: <integer>, <float>, or <string>
- name: <string>
  value: <integer, <float>, or <string>
... repeating as necessary
```

> Each element of the `parameterValues` list in the file must be either: 

Reviewer: ```diff
-Each element of the `parameterValues` list in the file must be either: 
+Each element of the `parameterValues` list in the file must be one of the following: 
```

1. A Job Parameter defined in the Job Bundle's Job Template; 

> 2. A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to; or 

Reviewer: I didn't know this! Perhaps we find a way to link out to one of the Queue Environment samples.

```diff
-A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to; or 
+A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to;
```

    * `deadline:maxFailedTasksCount` - Value must be an integer, and is passed as the `maxFailedTasksCount` request parameter to the API.
    * `deadline:maxRetriesPerTask` - Value must be an integer, and is passed as the `maxRetriesPerTask` request parameter to the API.

While a Job Template is always a template rather than a specific Job to run, the parameter values file allows
a Job Bundle to act as a template if some Job Parameters do not have values defined in this file, or as a specific job submission
if all Job Parameters have values.

For example, the [`blender_render` sample](https://github.com/aws-deadline/deadline-cloud-samples/tree/mainline/job_bundles/blender_render)
has no parameter values file, and its Job Template defines Job Parameters, like `BlenderSceneFile` and `OutputDir`, that
have no default values, so it must be used as a template from which Jobs are created. When we create a Job using this Job Bundle, with
`deadline bundle gui-submit blender_render/`, a new Job Bundle is written to the job history directory and that Job Bundle has

> a `parameter_values.yaml` file that contains the values of all parameters defined for the Job submission:

Reviewer: ```diff
-have no default values, so it must be used as a template from which Jobs are created. When we create a Job using this Job Bundle, with
+have no default values, so it must be used as a template from which Jobs are created. After we submit a Job using this Job Bundle
-`deadline bundle gui-submit blender_render/`, a new Job Bundle is written to the job history directory and that Job Bundle has
a `parameter_values.yaml` file that contains the values of all parameters defined for the Job submission:
+(with `deadline bundle gui-submit blender_render/` for example), a new Job Bundle is written to the job history directory and that Job Bundle has
-a `parameter_values.yaml` file that contains the values of all parameters defined for the Job submission:
+a `parameter_values.yaml` file that contains the values of all parameters we specified:
```

```bash
% cat ~/.deadline/job_history/\(default\)/2024-06/2024-06-20-01-JobBundle-Demo/parameter_values.yaml
parameterValues:
- name: deadline:targetTaskRunStatus
  value: READY
- name: deadline:maxFailedTasksCount
  value: 10
- name: deadline:maxRetriesPerTask
  value: 5
- name: deadline:priority
  value: 75
- name: BlenderSceneFile
  value: /private/tmp/bundle_demo/bmw27_cpu.blend
- name: Frames
  value: 1-10
- name: OutputDir
  value: /private/tmp/bundle_demo/output
- name: OutputPattern
  value: output_####
- name: Format
  value: PNG
- name: CondaPackages
  value: blender
- name: RezPackages
  value: blender
```

We could recreate the exact same Job by submitting this new Job Bundle instead with: 

```
deadline bundle submit ~/.deadline/job_history/\(default\)/2024-06/2024-06-20-01-JobBundle-Demo/
```

Note: The submitted Job Bundle is saved to your "job history directory" and the location of that directory 
can be found by running `deadline config get settings.job_history_dir`.

### Elements - Asset References

The `asset_references.json`/`asset_references.yaml` file in a Job Bundle is an interface for Deadline Cloud's

> Job Attachments feature that lists input files and directories, as well as output directories for the Jobs that you

Reviewer 3: Can we add a link for [Job Attachments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html#:~:text=Job%20attachments%20enable%20you%20to,bucket%20for%20your%20job%20attachments.)

Author: Of course!

submit. If you do not have all of your inputs and outputs for Job Attachments listed in this file
then you can still select them for your Job during submission with the `deadline bundle gui-submit` CLI command.

> This file has no effect if you are not using Deadline Cloud's Job Attachments feature.

Reviewer: Link out to some documentation about "Deadline Cloud's Job Attachments feature"
```
This file has no effect if you are not using Deadline Cloud's Job Attachments feature.
```

Reviewer: Might be worth mentioning that if paths in here aren't part of the Storage Profile (another link out), they will be uploaded to S3 using said Job Attachments feature

Reviewer 2: +1 on both these points

# AWS Deadline Cloud Job Bundles

> Job Bundles are one of the tools that you can use to define your Jobs for Amazon Deadline Cloud. They encapsulate

Reviewer 4: ```suggestion
Job Bundles are one of the tools that you can use to define your Jobs for AWS Deadline Cloud. They encapsulate
```

an [Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) into a directory
with additional information such the files and directories that your Jobs need for
Deadline Cloud's [Job Attachments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html)
feature. The [Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud) provides ways for you to use a Job Bundle
to submit Jobs to your Deadline Cloud Queues to be run.

> We recommend developing your own Job Bundle(s), either by-hand or programatically, for your custom Job submission needs. 

Reviewer 4: 'Recommend' makes it  sound like there is an alternative

Author: There is. You could develop just a Job Template and use the CreateJob API directly. I can clarify that point.

Reviewer 5: I think this paragraph would be clearer if it pointed out what bundles give you compared to job templates instead of giving a blanket recommendation. Something in the spirit of, "Job bundles build on job templates to deliver additional features like job attachments and a GUI submitter." Someone not needing these features (e.g. CMF plus in house job submission GUI) might be better off using CreateJob + a template.

Author: Good suggestion, Stephen, though that's covered by the first paragraph, no? I can do a bit better with this paragraph, though, I think.

where the contents of the Job-specific settings panel are automatically generated using the `userInterface` properties of Job Parameters
[defined in the Job Template](https://github.com/aws-deadline/deadline-cloud-samples/blob/bdd5ff5ea29eb7457c9a78ba39166b891b79151e/job_bundles/blender_render/template.yaml#L11-L19) within 
your Job Bundle, and the [Queue Environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html)
defined on the Queue that you are submitting to.

If you prefer a non-interactive submission workflow for your Job Bundles, then you can submit this Job Bundle with the command
`deadline bundle submit --yes --name Demo -p BlenderSceneFile=<location-of-your-scene-file> -p OutputDir=<file-path-for-job-outputs> blender_render/`
or use the `deadline.client.api.create_job_from_job_bundle` function in the [`deadline` Python package](https://github.com/aws-deadline/deadline-cloud).

All of the Job submitter plugins that have been developed by the AWS Deadline Cloud team, such as the 

> [Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your

Reviewer 6: According to the [Concepts and terminology for Deadline Cloud](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/concepts-terminology.html) topic of the AWS Deadline Cloud user guide, we should refer to these as "Deadline Cloud submitters" and not "submitter plugins" for consistency.

Job submission and then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) to
submit your Job to Deadline Cloud. You can see the Job Bundles that are

> submitted by looking in the job history directory on your workstation after submitting a Job using a submitter plugin or the Deadline

Reviewer 6: nit: how about "You can inspect the job bundles created for previously submitted jobs by looking in the job history directory"?

Cloud CLI. You can find your job history directory by running the command: `deadline config get settings.job_history_dir`.

## Elements of a Job Bundle

A Job Bundle is a directory structure that contains at least an
[Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) file, but may contain
other files as follows:

```
/template.yaml (or template.json)
/asset_references.yaml (or asset_references.json)
/parameter_values.yaml (or parameter_values.json)

> <plus any other Job-specific files that you'd like>

Reviewer 6: nit: just a little visual improvement with the dir tree lines
```suggestion
<BUNDLE_DIR>/
├── asset_references.yaml (or asset_references.json)
├── parameter_values.yaml (or parameter_values.json)
├── template.yaml\ (or template.json)
└── <other Job-specific files that you'd like>
```

```

The only required file is the Job Template (`template.yaml`/`template.json`) file that describes the structure and behaviour

> of your Job. The other files are optional, and are described in the following subsections.

Reviewer 6: nit: both required and optional files are described in the subsections, so how about:

```suggestion
The only required file is the Job Template (`template.yaml`/`template.json`) file that describes the structure and behaviour
of your Job. The files are described in the following subsections.
```

### Elements - Job Template

> The `template.yaml`/`template.json` file defines the runtime environment and all the processes that will run

Reviewer 4: this whole doc is a great start. It's not clear how deep we want to go but we're missing mentioning some key features like steps(barely mentioned), step dependencies, and job environments.

Author: Yes. I'm trying to keep this one narrowly focused on the delta between a Job Template and a Job Bundle. Otherwise, it can get quite out of hand if I try to explain everything -- I'd have to include a lot of details about path mapping, storage profiles, and job attachments when talking about asset references later. 

The plan is a series of documents that explain the various aspects of the system. One for Job Templates, one for Job Bundles (this one), and one for "assets" (path mapping, storage profiles, and job attachments).

> as part of an Amazon Deadline Cloud Job. It can be parameterized so that the same template can be used to

Reviewer 4: ```suggestion
as part of an AWS Deadline Cloud Job. It can be parameterized so that the same template can be used to
```

Job Bundle. The Job Bundle submission UI that is created will still allow values for these parameters to be modified.

The format of the file, in YAML, is:

```yaml
parameterValues:
- name: <string>
  value: <integer>, <float>, or <string>
- name: <string>
  value: <integer, <float>, or <string>
... repeating as necessary
```

Each element of the `parameterValues` list in the file must be one of the following: 

1. A Job Parameter defined in the Job Bundle's Job Template; 
2. A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to;
3. A special parameter that is passed to the [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html)
   when creating a Job with the Job Bundle:
    * `deadline:priority` - The value must be an integer, and is passed as the `priority` request parameter to the API.
    * `deadline:targetTaskRunStatus` - Value must be a string, and is passed as the `targetTaskRunStatus` request parameter to the API.
    * `deadline:maxFailedTasksCount` - Value must be an integer, and is passed as the `maxFailedTasksCount` request parameter to the API.

>     * `deadline:maxRetriesPerTask` - Value must be an integer, and is passed as the `maxRetriesPerTask` request parameter to the API.

Reviewer 4: is `deadline:storageProfileId` supported?

Author: [Not at this time.](https://github.com/aws-deadline/deadline-cloud/blob/c6334e37ca13f676ad8905a30f54c78d4cc081d5/src/deadline/client/ui/widgets/shared_job_settings_tab.py#L287-L304)

where the contents of the Job-specific settings panel are automatically generated using the `userInterface` properties of Job Parameters
[defined in the Job Template](https://github.com/aws-deadline/deadline-cloud-samples/blob/bdd5ff5ea29eb7457c9a78ba39166b891b79151e/job_bundles/blender_render/template.yaml#L11-L19) within 
your Job Bundle, and the [Queue Environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html)
defined on the Queue that you are submitting to.

If you prefer an alternative to a UI-based workflow for your Job Bundles, then you can submit this Job Bundle with the command
`deadline bundle submit --name Demo -p BlenderSceneFile=<location-of-your-scene-file> -p OutputDir=<file-path-for-job-outputs> blender_render/`
or use the `deadline.client.api.create_job_from_job_bundle` function in the [`deadline` Python package](https://github.com/aws-deadline/deadline-cloud).

All of the Deadline Cloud submitters that have been developed by the AWS Deadline Cloud team, such as the 

> [Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your

Reviewer 6: ```suggestion
[Autodesk Maya submitter](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your
```

or perhaps

```suggestion
submitter included in [Deadline Cloud for Maya](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your
```

Job submission and then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) to
submit your Job to Deadline Cloud. You can see the Job Bundles that are
submitted by looking in the job history directory on your workstation after submitting a Job using a submitter plugin or the Deadline
Cloud CLI. You can find your job history directory by running the command: `deadline config get settings.job_history_dir`.

## Elements of a Job Bundle

A Job Bundle is a directory structure that contains at least an
[Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) file, but may contain
other files such as:

```
/template.yaml (or template.json)
/asset_references.yaml (or asset_references.json)

> /parameter_values.yaml (or parameter_values.json)

Reviewer 5: I think it'd be worth making explicit somewhere that job bundles can be used both to define a reusable workflow (e.g. our sample bundles) and to bake/archive/freeze a particular job submission (e.g. `~/.deadline/job_history`). Someone might be confused what asset references are doing in a reusable job definition.

Author: That differentiation is made later on in the document.

langauge. 

For example, the [Job Template for the `blender_render` sample](https://github.com/aws-deadline/deadline-cloud-samples/blob/mainline/job_bundles/blender_render/template.yaml)
defines input parameters like `BlenderSceneFile` which is a file path:

```yaml
- name: BlenderSceneFile
  type: PATH
  objectType: FILE
  dataFlow: IN
  userInterface:
    control: CHOOSE_INPUT_FILE
    label: Blender Scene File
    groupLabel: Render Parameters
    fileFilters:
    - label: Blender Scene Files
      patterns: ["*.blend"]
    - label: All Files
      patterns: ["*"]
  description: >
    Choose the Blender scene file you want to render. Use the 'Job Attachments' tab
    to add textures and other files that the job needs.
```

`userInterface`, `objectType`, and `dataFlow` are optional Job Parameter properties that the Deadline Cloud CLI 
understands when present in a Job Template within a Job Bundle. 

`userInterface` properties control the behaviour and appearance of fields in automatically generated Job submission UIs; both via 
the `deadline bundle gui-submit` command line, and within Deadline Cloud submitters for applications

> such as the [Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya). 

Reviewer 6: ```suggestion
such as the [Autodesk Maya submitter](https://github.com/aws-deadline/deadline-cloud-for-maya). 
```

or perhaps:

```suggestion
submitter included in [Deadline Cloud for Maya](https://github.com/aws-deadline/deadline-cloud-for-maya).
```

Additional samples of Job Templates can be found in all of the Job Bundles in this repository as well as in the 
[samples provided](https://github.com/OpenJobDescription/openjd-specifications/tree/mainline/samples)
by Open Job Description. Please see https://github.com/OpenJobDescription/openjd-specifications/wiki
for more information on Job Templates.

### Elements - Parameter Values

The `parameter_values.yaml`/`parameter_values.json` file in a Job Bundle gives you a place to "bake"
the values of some of the Job Parameters, and/or [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html)
request arguments into the Job Bundle so that the values do not have to be given when submitting a Job using the
Job Bundle. The Job Bundle submission UI that is created will still allow values for these parameters to be modified.

The format of the file, in YAML, is:

```yaml
parameterValues:
- name: <string>
  value: <integer>, <float>, or <string>
- name: <string>
  value: <integer, <float>, or <string>
... repeating as necessary
```

> Each element of the `parameterValues` list in the file must be one of the following: 

Reviewer 6: Maybe we change "must" to "can". Strictly speaking, these values are ignored if they are not any of the below. The reason being that a parameter value was baked into a job bundle based on the queue's job parameters at the time of submission. When submitting to a different queue without the parameter, or if the queue's environments are modified/removed and the parameter is no longer defined, the parameter is ignored.

1. A Job Parameter defined in the Job Bundle's Job Template; 
2. A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to;
3. A special parameter that is passed to the [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html)
   when creating a Job with the Job Bundle:
    * `deadline:priority` - The value must be an integer, and is passed as the `priority` request parameter to the API.
    * `deadline:targetTaskRunStatus` - Value must be a string, and is passed as the `targetTaskRunStatus` request parameter to the API.
    * `deadline:maxFailedTasksCount` - Value must be an integer, and is passed as the `maxFailedTasksCount` request parameter to the API.
    * `deadline:maxRetriesPerTask` - Value must be an integer, and is passed as the `maxRetriesPerTask` request parameter to the API.

While a Job Template is always a template rather than a specific Job to run, the parameter values file allows
a Job Bundle to act as a template if some Job Parameters do not have values defined in this file, or as a specific job submission

> if all Job Parameters have values.

Reviewer 6: The parameter values in a job bundle can be overridden, so maybe a better mental model is "a specific job submission with parameters that can be overridden".

Author: It's not a specific submission unless all of the Job Parameters either have defaults or have values defined in parameter_values. So, I don't think it's as simple as you've stated.

Reviewer 6: I suppose a job template could also be considered a specific job submission under the same condition (all of the parameters in the template have defaults), right?

What does the parameter values file add beyond using a job template and modifying the defaults? I think the concept we're trying to document here is feature/behavior of the `deadline-cloud` library when it submits a job bundle. It persists a copy of the job bundle with the chosen job parameter values baked into the parameter values file. In this way, the job bundles in the job history directory act as a job submission history. Those bundles can be submitted using the baked parameter values to repeat the job submission.

Author: > What does the parameter values file add beyond using a job template and modifying the defaults?

That and providing values for a small handful of the CreateJob API parameters.

 What I'm trying to convey here is that if you have a parameter_values file and that file has values for all non-defaulted job parameters, then the Bundle is, in effect, a representation of a specific concrete Job rather than a template from which Jobs can be created.

And, I think, with that I see a better way to word this paragraph.

Reviewer 6: There was an offline discussion where we discussed that job bundle parameter values do have the same behavior of acting as a default value when submitting a bundle and that they can be overridden.

The differences above and beyond a job template defaults are:
1.  Job bundle parameter values accept wider ranges of values, such as:
    1.  Allow specifying values for `PATH` parameters outside the job bundle directory
    2.  Allow specifying `deadline:*` values used in the `CreateJob` Deadline Cloud API requests.
2. A much more simple way for tooling/automation to provide new/modified default values for a job template: a simple JSON file with name/value pairs

We agreed that this `README.md` is intended to be a high-level informal introduction to job bundles. We can adjust the documentation in the future to help clarify to readers that parameter values can be overridden and are still allow providing different values when submitting a job from the bundle.

# AWS Deadline Cloud Job Bundles

> Job Bundles are one of the tools that you can use to define your Jobs for AWS Deadline Cloud. They encapsulate

Reviewer 6: Great doc! Just one comment here for a general suggestion that we should not capitalize terms like "jobs", "job templates", "queues", etc. to be consistent with the writing style of the user guide.

an [Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) into a directory
with additional information such the files and directories that your Jobs need for
Deadline Cloud's [Job Attachments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html)
feature. The [Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud) provides ways for you to use a Job Bundle
to submit Jobs to your Deadline Cloud Queues to be run.

We recommend developing your own Job Bundle(s), either by-hand or programatically, for your custom Job submission needs
rather than the alternative of developing solely an Open Job Description Job Template and using the 

> [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html) API directly. 

Reviewer 7: We say API twice

and an intuitive graphical Job submission interface. For example, after 
[configuring the Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud/blob/mainline/README.md#configuration) 
you can run it with the Blender sample in this samples directory (`deadline bundle gui-submit blender_render/`) to see:

![UI Shared Settings](../.images/blender_submit_shared_settings.png) ![UI Job Settings](../.images/blender_submit_job_settings.png) ![UI Job Attachments](../.images/blender_submit_job_attachments.png) 

where the contents of the Job-specific settings panel are automatically generated using the `userInterface` properties of Job Parameters
[defined in the Job Template](https://github.com/aws-deadline/deadline-cloud-samples/blob/bdd5ff5ea29eb7457c9a78ba39166b891b79151e/job_bundles/blender_render/template.yaml#L11-L19) within 
your Job Bundle, and the [Queue Environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html)
defined on the Queue that you are submitting to.

If you prefer a non-interactive submission workflow for your Job Bundles, then you can submit this Job Bundle with the command

> `deadline bundle submit --yes --name Demo -p BlenderSceneFile=<location-of-your-scene-file> -p OutputDir=<file-path-for-job-outputs> blender_render/`

Reviewer 7: We should remove the `--yes` here, it looks out of place and we shouldn't normalize the usage of it, as it's a security precaution to ensure submissions are being made from correct locations.

or use the `deadline.client.api.create_job_from_job_bundle` function in the [`deadline` Python package](https://github.com/aws-deadline/deadline-cloud).

All of the Job submitter plugins that have been developed by the AWS Deadline Cloud team, such as the 
[Autodesk Maya plugin](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your
Job submission and then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) to
submit your Job to Deadline Cloud. You can see the Job Bundles that are
submitted by looking in the job history directory on your workstation after submitting a Job using a submitter plugin or the Deadline
Cloud CLI. You can find your job history directory by running the command: `deadline config get settings.job_history_dir`.

## Elements of a Job Bundle

A Job Bundle is a directory structure that contains at least an

> [Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) file, but may contain

Reviewer 6: 1.  Should this link to OpenJD should be moved to the job template section below to be closer in proximity to the place where readers want to dive in?
2.  I wonder if we should link straight to the template spec instead of the wiki landing page.

    I imagine the intent behind linking to the wiki landing page was that we will have multiple spec versions over time. Unfortunately, the landing page does not have the best linkage to the current template specification so readers have to dig to get to the relevant spec reference information they need. Maybe a better approach is to improve the OpenJD wiki navigation to have more prominent links to the template spec(s) on the landing page.

Author: I see what you're saying with the location of the link, but I think that the flow is better with it here.

The job template information is pretty easy to navigate to in the sidebar on the wiki, and I think that the wiki landing page is a nicer introduction than plopping the reader directly into the formalized job template specification.

Reviewer 6: > The job template information is pretty easy to navigate to in the sidebar on the wiki

It is there, but I'm not sure it is prominent or it is clear that this where you should go for help writing job templates. How about we keep this link to point to the landing page, but submit a future PR to improve the landing page to help direct users to the job template schema?

Author: I'm currently working on material for that wiki regarding authoring job templates. Improvements are coming. :-)

Reviewer 6: Sounds good. I'll resolve this

> other files as follows:

Reviewer 7: ```suggestion
other files such as:
```

Nit: since this list is not exhaustive nor mandatory, I suggest changing the phrasing from 'as follows' to 'such as'

```
/template.yaml (or template.json)
/asset_references.yaml (or asset_references.json)
/parameter_values.yaml (or parameter_values.json)
<plus any other Job-specific files that you'd like>
```

The only required file is the Job Template (`template.yaml`/`template.json`) file that describes the structure and behaviour
of your Job. The other files are optional, and are described in the following subsections.

### Elements - Job Template

> The `template.yaml`/`template.json` file defines the runtime environment and all the processes that will run

Reviewer 6: > ... **all** the processes that will run as part of an AWS Deadline Cloud Job

Queue environments also define processes that run as part of the job. I wonder if we should incorporate this. 

Maybe:

>   The `template.yaml`/`template.json` file define the runtime environment and process that will run as part of an AWS Deadline Cloud job.

We could have an info box below that adds the additional queue env detail:

```md
---
**NOTE:** You can also setup [OpenJobDescription external environments](https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#12-environment-template) to Amazon Deadline Cloud queues (see [creating queue environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html) in the Amazon Deadline Cloud user guide). Queue environments combine with the job template `template.yaml` / `template.json` to define the job processes.
---
```

What do you think?

Author: Good suggestion

> The `template.yaml`/`template.json` file defines the runtime environment and all the processes that will run

Reviewer 7: ```suggestion
The Job Template file defines the runtime environment and all the processes that will run
```
We just did the template.json/template.yaml thing one line above, so it seems redundant to repeat it here.

as part of an AWS Deadline Cloud Job. It can be parameterized so that the same template can be used to
create Jobs that differ only in their input values; much like a function or template in your favourite programming
langauge. 

For example, the [Job Template for the `blender_render` sample](https://github.com/aws-deadline/deadline-cloud-samples/blob/mainline/job_bundles/blender_render/template.yaml)
defines input parameters like `BlenderSceneFile` which is a file path:

```yaml
- name: BlenderSceneFile
  type: PATH
  objectType: FILE
  dataFlow: IN
  userInterface:
    control: CHOOSE_INPUT_FILE
    label: Blender Scene File
    groupLabel: Render Parameters
    fileFilters:
    - label: Blender Scene Files
      patterns: ["*.blend"]
    - label: All Files
      patterns: ["*"]
  description: >
    Choose the Blender scene file you want to render. Use the 'Job Attachments' tab
    to add textures and other files that the job needs.
```

> `userInterface`, `objectType`, and `dataFlow` are optional Job Parameter properties used by the Deadline Cloud CLI when

Reviewer 7: Are they really _used_ the CLI? Is there a better way to describe how optional parameter values are utilized by the system overall?

present in a Job Template within a Job Bundle. 

> The `userInterface` property control the behaviour of automatically generated Job submission UIs; both via 

Reviewer 7: ```suggestion
`userInterface` properties control the behaviour and apperance of fields in automatically generated Job submission UIs; both via 
```

        # inputs to run the Job.
        filenames:
        - <list of file paths>
        # Directories on the submitting workstation whose contents are needed as inputs
        # to run the Job.
        directories:
        - <list of directory paths>

    outputs:
        # Directories on the submitting workstation where the Job would write output files
        # if it was running locally.
        directories:
        - <list of directory paths>

    # Paths that are referenced by the job, but not necessarily input or output.
    # Use this if your job uses the name of a path in some way, but does not explicitly need
    # the contents of that path.
    referencedPaths:
    - <list of directory paths>
```

When selecting which input or output files to upload to [Amazon S3](https://aws.amazon.com/s3/), the Job Attachments
feature compares the file path against the paths listed in your
[Storage Profiles](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-shared.html). Each Storage Profile
abstracts a network fileshare on your network that is mounted on your workstations and worker hosts, so Job Attachments

> only uploads files to S3 if the file is not contained on one of these file shares.

Reviewer 8: Small detail to adjust here - this only applies to SHARED storage profile file system locations. Files in LOCAL file system locations will be attached to the job.

# AWS Deadline Cloud Job Bundles

> Job Bundles are one of the tools that you can use to define your Jobs for AWS Deadline Cloud. They encapsulate

Reviewer 8: How would you feel about making the opening sentence more opinionated? E.g. "Job Bundles are the easiest way to define your jobs for AWS Deadline Cloud." The way it's written, it sounds like there are a bunch of different ways to do it, and you should consider all those different ways, but I think it's better to start with just job bundles.

Author: I can get behind that, sure.

an [Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) into a directory
with additional information such the files and directories that your Jobs need for
Deadline Cloud's [Job Attachments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html)
feature. The [Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud) provides ways for you to use a Job Bundle
to submit Jobs to your Deadline Cloud Queues to be run.

We recommend developing your own Job Bundle(s), either by-hand or programatically, to benefit from their

> integration with Job Attachments and an intuitive graphical Job submission interface. For example, after 

Reviewer 8: This will be more clear without the job attachments reference. When you write a job bundle, you can structure it to work with both job attachments and shared file systems, and someone reading this might interpret it to mean that job bundles are for job attachments.

Author: Sure, we can try that on. There's already a call-out to the integration with Job Attachments in the intro paragraph.

[configuring the Deadline Cloud CLI](https://github.com/aws-deadline/deadline-cloud/blob/mainline/README.md#configuration) 
you can run it with the Blender sample in this samples directory (`deadline bundle gui-submit blender_render/`) to see:

![UI Shared Settings](../.images/blender_submit_shared_settings.png) ![UI Job Settings](../.images/blender_submit_job_settings.png) ![UI Job Attachments](../.images/blender_submit_job_attachments.png) 

where the contents of the Job-specific settings panel are automatically generated using the `userInterface` properties of Job Parameters
[defined in the Job Template](https://github.com/aws-deadline/deadline-cloud-samples/blob/bdd5ff5ea29eb7457c9a78ba39166b891b79151e/job_bundles/blender_render/template.yaml#L11-L19) within 
your Job Bundle, and the [Queue Environments](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/create-queue-environment.html)

> defined on the Queue that you are submitting to.

Reviewer 8: A nitpick here: The queue environment parameters are shown in the shared job settings tab, but this says that they're shown in the job-specific settings tab. I suspect that the queue you submitted to has no queue environments with parameters, maybe it would be good to add one and re-take the first screenshot?

If you prefer an alternative to a UI-based workflow for your Job Bundles, then you can submit this Job Bundle with the command
`deadline bundle submit --name Demo -p BlenderSceneFile=<location-of-your-scene-file> -p OutputDir=<file-path-for-job-outputs> blender_render/`
or use the `deadline.client.api.create_job_from_job_bundle` function in the [`deadline` Python package](https://github.com/aws-deadline/deadline-cloud).

> You can also develop your Job as a Job Template and use the

Reviewer 8: I think it's better to exclude this sentence. For example, things like passing job parameters and likely other things are trickier this way. We can create an advanced documentation page that covers all those details later.

Author: This paragraph is a breadcrumb for pipeline developers in response to a specific question that we had from a customer. If/when we create such an advanced page then I think it's reasonable to replace this with a link to that, but I'm inclined to leave it in the meantime.

to make use of a Job Bundle's integration with the Job Attachments feature.

All of the Deadline Cloud submitters that have been developed by the AWS Deadline Cloud team, such as the 
[Autodesk Maya submitter](https://github.com/aws-deadline/deadline-cloud-for-maya), are simply generating a Job Bundle for your
Job submission and then using the [Deadline Cloud Python package](https://github.com/aws-deadline/deadline-cloud) to
submit your Job to Deadline Cloud. You can inspect the job bundles created for previously submitted jobs by looking in the job history directory.
You can find your job history directory by running the command: `deadline config get settings.job_history_dir`.

## Elements of a Job Bundle

A Job Bundle is a directory structure that contains at least an
[Open Job Description Job Template](https://github.com/OpenJobDescription/openjd-specifications/wiki) file, but may contain
other files such as:

```
<BUNDLE_DIR>/
├── asset_references.yaml (or asset_references.json)
├── parameter_values.yaml (or parameter_values.json)
├── template.yaml (or template.json)

> └── <other Job-specific files that you'd like>

Reviewer 8: "files and directories"

```

The only required file is the Job Template (`template.yaml`/`template.json`) file that describes the structure and behaviour
of your Job. The files are described in the following subsections.

### Elements - Job Template

The Job Template file defines the runtime environment and the processes that will run
as part of an AWS Deadline Cloud Job. It can be parameterized so that the same template can be used to
create Jobs that differ only in their input values; much like a function or template in your favourite programming

> langauge. 

Reviewer 8: typo: language

> > **NOTE:** You can also setup [Open Job Description external environments](https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#12-environment-template)

Reviewer 8: This note feels out of place here. Maybe move this to the intro paragraph as a sentence like "When you submit a job to Deadline Cloud, it gets run within any queue environments that are applied to the queue. Queue environments use the [Open Job Description external environment specification](...).

    to add textures and other files that the job needs.
```

`userInterface`, `objectType`, and `dataFlow` are optional Job Parameter properties that the Deadline Cloud CLI 
understands when present in a Job Template within a Job Bundle. 

`userInterface` properties control the behaviour and appearance of fields in automatically generated Job submission UIs; both via 
the `deadline bundle gui-submit` command line, and within Deadline Cloud submitters for applications
such as the [Autodesk Maya submitter](https://github.com/aws-deadline/deadline-cloud-for-maya). 
In this example, the UI widget for inputting a value for `BlenderSceneFile` will be a file-selection dialog that
allows filtering to see only Blender's `.blend` files or all files, and within a widget group called "Render Parameters":

![BlenderSceneFile UI Widget](../.images/blender_submit_scene_file_widget.png)

See the [gui_control_showcase sample](https://github.com/aws-deadline/deadline-cloud-samples/tree/mainline/job_bundles/gui_control_showcase)
for additional example uses of the `userInterface` property.

The `objectType` and `dataFlow` properties control the behaviour of Deadline Cloud's 

> [Job Attachments feature](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html) when submitting

Reviewer 8: While this is true, could we start by describing the semantics the properties have, and then explain that job attachments uses those semantics? We want other systems to use those semantics as well, and those other systems aren't necessarily job attachments.

Author: Good suggestion. I've reworded this paragraph.

The format of the file, in YAML, is:

```yaml
parameterValues:
- name: <string>
  value: <integer>, <float>, or <string>
- name: <string>
  value: <integer, <float>, or <string>
... repeating as necessary
```

Each element of the `parameterValues` list in the file can be one of the following: 

1. A Job Parameter defined in the Job Bundle's Job Template; 
2. A Job Parameter defined in a Queue Environment on the Queue that you are submitting the Job to;
3. A special parameter that is passed to the [deadline:CreateJob API](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_CreateJob.html)
   when creating a Job with the Job Bundle:
    * `deadline:priority` - The value must be an integer, and is passed as the `priority` request parameter to the API.
    * `deadline:targetTaskRunStatus` - Value must be a string, and is passed as the `targetTaskRunStatus` request parameter to the API.
    * `deadline:maxFailedTasksCount` - Value must be an integer, and is passed as the `maxFailedTasksCount` request parameter to the API.
    * `deadline:maxRetriesPerTask` - Value must be an integer, and is passed as the `maxRetriesPerTask` request parameter to the API.

> While a standalone Job Template is viewed as a template from which Jobs can be created, a Job Bundle can be viewed as both

Reviewer 8: A standalone template could be viewed as either one as well. E.g. a template that takes no parameters will define one specific job. I'd suggest to remove the contrast here, and just talk about the job bundle this way. "A job bundle can function as either a template from which to create jobs, or as a single specific job."

Reviewer 6: See https://github.com/aws-deadline/deadline-cloud-samples/pull/13#discussion_r1663002017. I felt the same way. Even a job bundle with parameter values for all parameters can be used like a template by providing overridden values.

a `parameter_values.yaml` file that contains the values of all parameters we specified:

```bash
% cat ~/.deadline/job_history/\(default\)/2024-06/2024-06-20-01-JobBundle-Demo/parameter_values.yaml
parameterValues:
- name: deadline:targetTaskRunStatus
  value: READY
- name: deadline:maxFailedTasksCount
  value: 10
- name: deadline:maxRetriesPerTask
  value: 5
- name: deadline:priority
  value: 75
- name: BlenderSceneFile
  value: /private/tmp/bundle_demo/bmw27_cpu.blend
- name: Frames
  value: 1-10
- name: OutputDir
  value: /private/tmp/bundle_demo/output
- name: OutputPattern
  value: output_####
- name: Format
  value: PNG
- name: CondaPackages
  value: blender
- name: RezPackages
  value: blender
```

We could recreate the exact same Job by submitting this new Job Bundle instead with: 

```
deadline bundle submit ~/.deadline/job_history/\(default\)/2024-06/2024-06-20-01-JobBundle-Demo/
```

Note: The submitted Job Bundle is saved to your "job history directory" and the location of that directory 
can be found by running `deadline config get settings.job_history_dir`.

### Elements - Asset References

> The `asset_references.yaml`/`asset_references.json` file in a Job Bundle is an interface for Deadline Cloud's

Reviewer 8: I think it's better to first state that it defines the input and output files that the job will access when it runs, and then state that Deadline Cloud's job attachments feature uses that to determine what to upload and download. Then it's clear immediately that a developer can use this for other purposes than job attachments.

[Job Attachments feature](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/storage-job-attachments.html)
that lists input files and directories, as well as output directories for the Jobs that you
submit. If you do not have all of your inputs and outputs for Job Attachments listed in this file
then you can still select them for your Job during submission with the `deadline bundle gui-submit` CLI command.

> This file has no effect if you are not using Deadline Cloud's Job Attachments feature.

Reviewer 8: I don't think we should state this, because it will stop being true the moment someone implements an Open Job Description tool that uses the file outside of job attachments.

Author: Sure
